src/crypto/CHIPCryptoPAL.h

```diff
@@ -118,17 +148,99 @@ class P256PrivateKey : public ECPKey
 {
 public:
     SupportedECPKeyTypes Type() const override { return SupportedECPKeyTypes::ECP256R1; }
     size_t Length() const override { return kP256_PublicKey_Length; }
     operator uint8_t *() const override { return (uint8_t *) bytes; }
 
+    CHIP_ERROR ECDSA_validate_msg_signature(const uint8_t * msg, const size_t msg_length,
+                                            const P256ECDSASignature & signature) const override;
+
 private:
     uint8_t bytes[kP256_PublicKey_Length];
 };
 
+template <typename PK, typename Secret, typename Sig>
+class ECPKeypair
+{
+public:
+    /** @brief Generate a new Certificate Signing Request (CSR).
+     * @param csr Newly generated CSR
+     * @param csr_length The caller provides the length of input buffer (csr). The function returns the actual length of generated
+     *CSR.
+     * @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
+     **/
+    virtual CHIP_ERROR NewCertificateSigningRequest(uint8_t * csr, size_t & csr_length) = 0;
+
+    /**
+     * @brief A function to sign a msg using ECDSA
+     * @param msg Message that needs to be signed
+     * @param msg_length Length of message
+     * @param out_signature Buffer that will hold the output signature. The signature consists of: 2 EC elements (r and s),
+     *represented as ASN.1 DER integers, plus the ASN.1 sequence Header
+     * @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
+     **/
+    virtual CHIP_ERROR ECDSA_sign_msg(const uint8_t * msg, const size_t msg_length, Sig & out_signature) = 0;
+
+    /** @brief A function to derive a shared secret using ECDH
+     * @param remote_public_key Public key of remote peer with which we are trying to establish secure channel. remote_public_key is
+     * ASN.1 DER encoded as padded big-endian field elements as described in SEC 1: Elliptic Curve Cryptography
+     * [https://www.secg.org/sec1-v2.pdf]
+     * @param out_secret Buffer to write out secret into. This is a byte array representing the x coordinate of the shared secret.
+     * @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
+     **/
+    virtual CHIP_ERROR ECDH_derive_secret(const PK & remote_public_key, Secret & out_secret) const = 0;
+
+    virtual const PK & Pubkey() = 0;
+};
+
+class P256Keypair : public ECPKeypair<P256PublicKey, P256ECDHDerivedSecret, P256ECDSASignature>
+{
+public:
+    /** @brief Initialize the keypair.
+     * @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
+     **/
+    CHIP_ERROR Initialize();
+
+    /** @brief Generate a new Certificate Signing Request (CSR).
+     * @param csr Newly generated CSR
+     * @param csr_length The caller provides the length of input buffer (csr). The function returns the actual length of generated
+     *CSR.
+     * @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
+     **/
+    virtual CHIP_ERROR NewCertificateSigningRequest(uint8_t * csr, size_t & csr_length) override;
```

Comment:
nit: virtual should not be needed when 'override' is used. override is only used on virtuals.